cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -411,6 +433,34 @@ public boolean updateAppAllowed(ApplicationId appId) throws Exception {
     return version.endsWith(ApplicationId.DEFAULT_VERSION);
   }
 
+  /**
+   * To determine whether the app is allowed to be deployed:
+   * Do not deploy when the parent version is not the latest.
+   *
+   * @param appId the id of the application to be determined
+   * @param parentVersion the version of the application from which the app is deployed
+   * @return whether the app version is allowed to be deployed
+   */
+  public boolean deployAppAllowed(ApplicationId appId, String parentVersion) throws Exception {
+    accessEnforcer.enforce(appId, authenticationContext.getPrincipal(), StandardPermission.GET);
+    ApplicationMeta latest = store.getLatestAppVersion(appId);
+    // App does not exist
+    if (parentVersion == null || parentVersion.isEmpty() || latest == null || latest.getSpec() == null) {
+      //parent version in request is null then Allow to create a new one (Cannot create an app
+      // from a non existing parent version)
+      return parentVersion == null || parentVersion.isEmpty();
+    }
+    ApplicationSpecification parentAppSpec = store.getApplication(new ApplicationId(appId.getNamespace(),
```

Comment:
You don't need to get the parentAppSpec using parentVersion, user should only be allowed to deploy the app if parentVersion is latest version, can just remove line 453 to 458